### PR TITLE
Add bailout case to sourcemap searching code

### LINKED ISSUE #25209: Go to definition takes an unreasonably long time in a blank UWP app
1. Create a UWP app in Visual Studio 2017 (comes with a 100K line generated winrtrefs.d.ts file)
2. Type something that references the Windows APIs (I've been typing `Windows.Perception.People.HeadPose`, but any other type will work) and hit F12

EXPECTED:
Operation completes reasonably quickly (under 10 seconds?)

ACTUAL:
Took over 7 minutes on my machine

[This](https://github.com/Microsoft/TypeScript/blob/master/src/services/services.ts#L1521) is the offending loop.

The file is massive, but in TypeScript 2.8 the same operation only took a second or two. Go to definition should scale gracefully.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #25209

The test I added (a namespace with 580k blank lines inside it) drops from 21s to 1s with this change, so I think this aughta solve the problem.

cc @rbuckton who's more familiar with the performance aspects of v8's implementation, since optimizing this, even though it's only run once per file (per edit, potentially, since files are discarded on edit), is probably a good idea. It's be nice if you could remark on this style of matching - I don't think string slicing is worse than the `gm` regex we were using before, but I'm unsure, personally (I think it's better with the bail case we have now?).

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -1519,11 +1520,15 @@ namespace ts {
...
             for (let index = starts.length - 1; index >= 0; index--) {
-                sourceMapCommentRegExp.lastIndex = starts[index];
-                const comment = sourceMapCommentRegExp.exec(mappedFile.text);
+                const lineText = mappedFile.text.substring(starts[index], starts[index + 1]);
+                const comment = sourceMapCommentRegExp.exec(lineText);
                 if (comment) {
                     return comment[1];
                 }
+                // If we see a nonwhitespace line, break, to avoid scanning up the entire file
+                else if (!lineText.match(whitespaceRegExp)) {
```
Comment (rbuckton) [inline src/services/services.ts]: The only other comment allowed after `//# sourceMappingURL` is `//# sourceURL`, which also needs to be skipped.

**Comment by minestarks** [comment]
I'm questioning the value of checking in the massive test file. It's not more likely to fail than a small test file. If performance regresses, it'll just take longer to pass. Which is not a test failure anyone's likely to notice.

Other than that -- I've confirmed these changes fix #25209 and they look fine to me.

**Comment by weswigham** [comment]
> Which is not a situation that anyone's likely to notice.

20s longer of tests _usually_ get noticed. Usually. At least they have in the past, since the test duration metrics are recorded locally, and that duration is amplified when testing on less powerful machines. (And if anything the git history will be useful if it regresses and someone notices only by how their tests have started to take significantly longer) Unfortunately, I don't think I can actually make it large enough to crash or otherwise error (without crashing both before/after for other reasons anyway, like the file being way too large to fit in memory), and we don't have any test logic for fail-on-too-slow (which is always hard to set a bar for when test are running on multiple machines with different performance profiles). There is a test timeout of 40s... I could try to make it large enough to trigger the timeout in a pre-fixed situation? It probably just means doubling the length of the file to 1m lines.

**Comment by minestarks** [comment]
Usually performance tests are separate from functional tests, because of the reasons you pointed out. These two kinds of tests have fundamentally different definitions of pass and fail.

I'm not too familiar with how we structure tests in this repo. If there's precedence for this, and you believe it adds value, then OK by me.

Thanks for the fix!

**Pull request merged**